sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMultiFileReader.scala

```diff
@@ -0,0 +1,251 @@
+   * @param conf
+   * @param filters push down filters
+   * @return Callable[HostMemoryBuffersWithMetaDataBase]
+   */
+  def getBatchRunner(
+    file: PartitionedFile,
+    conf: Configuration,
+    filters: Array[Filter]): Callable[HostMemoryBuffersWithMetaDataBase]
+
+  /**
+   * decode HostMemoryBuffers by GPU
+   * @param fileBufsAndMeta
+   * @return
+   */
+  def readBatch(
+    fileBufsAndMeta: HostMemoryBuffersWithMetaDataBase): Option[ColumnarBatch]
+
+  def getLogTag: String
+
+  override def next(): Boolean = {
+    withResource(new NvtxRange(getLogTag + " readBatch", NvtxColor.GREEN)) { _ =>
+      if (isInitted == false) {
+        initAndStartReaders()
+      }
+      batch.foreach(_.close())
+      batch = None
+      // if we have batch left from the last file read return it
+      if (currentFileHostBuffers.isDefined) {
+        if (getSizeOfHostBuffers(currentFileHostBuffers.get) == 0) {
+          next()
+        }
+        batch = readBatch(currentFileHostBuffers.get)
+      } else {
+        currentFileHostBuffers = None
+        if (filesToRead > 0 && !isDone) {
+          val fileBufsAndMeta = tasks.poll.get()
+          filesToRead -= 1
+          TrampolineUtil.incBytesRead(inputMetrics, fileBufsAndMeta.bytesRead)
+          InputFileUtils.setInputFileBlock(
+            fileBufsAndMeta.partitionedFile.filePath,
+            fileBufsAndMeta.partitionedFile.start,
+            fileBufsAndMeta.partitionedFile.length)
+
+          if (getSizeOfHostBuffers(fileBufsAndMeta) == 0) {
+            // if sizes are 0 means no rows and no data so skip to next file
+            // file data was empty so submit another task if any were waiting
+            addNextTaskIfNeeded()
+            next()
+          } else {
+            batch = readBatch(fileBufsAndMeta)
+            // the data is copied to GPU so submit another task if we were limited
+            addNextTaskIfNeeded()
+          }
+        } else {
+          isDone = true
+          metrics(PEAK_DEVICE_MEMORY) += maxDeviceMemory
+        }
+      }
+    }
+
+    // this shouldn't happen but if somehow the batch is None and we still
+    // have work left skip to the next file
+    if (!batch.isDefined && filesToRead > 0 && !isDone) {
+      next()
+    }
+
+    // This is odd, but some operators return data even when there is no input so we need to
+    // be sure that we grab the GPU
+    GpuSemaphore.acquireIfNecessary(TaskContext.get())
+    batch.isDefined
+  }
+
+  override def get(): ColumnarBatch = {
+    val ret = batch.getOrElse(throw new NoSuchElementException)
+    batch = None
+    ret
+  }
+
+  private def getSizeOfHostBuffers(fileInfo: HostMemoryBuffersWithMetaDataBase): Long = {
+    fileInfo.memBuffersAndSizes.map(_._2).sum
+  }
+
+  private def addNextTaskIfNeeded(): Unit = {
+    if (tasksToRun.size > 0 && !isDone) {
+      val runner = tasksToRun.dequeue()
+      tasks.add(MultiFileThreadPoolFactory.submitToThreadPool(runner, numThreads))
```

Comment:
this seems odd in abstract code to use the class in parquet file. I think we discussed having a thread pool for each file type so we should abstract this out to call a method that does the right thing based on reader type.

Reply:
Yes, I will fix this issue.